web/pf4/src/components/ChartWithLegend.tsx

```diff
@@ -0,0 +1,231 @@
+
+type Props = {
+  data: VCLines;
+  seriesComponent: any;
+  unit: string;
+  chartHeight?: number;
+  groupOffset?: number;
+  fill?: boolean;
+  stroke?: boolean;
+  moreChartProps?: ChartProps;
+  overlay?: VCOverlay;
+};
+
+type State = {
+  width: number;
+  hiddenSeries: Set<number>;
+};
+
+class ChartWithLegend extends React.Component<Props, State> {
+  containerRef: React.RefObject<HTMLDivElement>;
+
+  constructor(props: Props) {
+    super(props);
+    this.containerRef = React.createRef<HTMLDivElement>();
+    this.state = { width: 0, hiddenSeries: new Set() };
+  }
+
+  handleResize = () => {
+    if (this.containerRef && this.containerRef.current) {
+      this.setState({ width: this.containerRef.current.clientWidth });
+    }
+  };
+
+  componentDidMount() {
+    setTimeout(() => {
+      this.handleResize();
+      window.addEventListener('resize', this.handleResize);
+    });
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.handleResize);
+  }
+
+  render() {
+    const scaleInfo = this.scaledAxisInfo(this.props.data);
+    const groupOffset = this.props.groupOffset || 0;
+
+    const dataWithOverlay = this.props.overlay ? this.props.data.concat(this.props.overlay.data) : this.props.data;
+    const overlayIdx = this.props.data.length;
+    const showOverlay = this.props.overlay && !this.state.hiddenSeries.has(overlayIdx);
+    const overlayRightPadding = showOverlay ? 30 : 0;
+
+    const legend = buildLegendInfo(dataWithOverlay, this.state.width);
+    const height = 300 + legend.height;
+    const padding = { top: 10, bottom: 20, left: 40, right: 10 + overlayRightPadding };
+    padding.bottom += legend.height;
+
+    const events = this.props.data.map((_, idx) => this.registerEvents(idx, 'serie-' + idx));
+    let overlayFactor = 1.0;
+    if (this.props.overlay) {
+      events.push(this.registerEvents(overlayIdx, 'overlay'));
+      // Normalization for y-axis display to match y-axis domain of the main data
+      // (see https://formidable.com/open-source/victory/gallery/multiple-dependent-axes/)
+      const mainMax = Math.max(...this.props.data.map(line => Math.max(...line.datapoints.map(d => d.y))));
+      const overlayMax = Math.max(...this.props.overlay.data.datapoints.map(d => d.y));
+      if (overlayMax !== 0) {
+        overlayFactor = mainMax / overlayMax;
+      }
+    }
+
+    return (
+      <div ref={this.containerRef}>
+        <Chart
+          height={height}
+          width={this.state.width}
+          padding={padding}
+          events={events}
+          containerComponent={createContainer()}
+          scale={{x: 'time'}}
+          {...this.props.moreChartProps}
+        >
+          <ChartGroup offset={groupOffset}>
+            {this.props.data.map((serie, idx) => {
+              if (this.state.hiddenSeries.has(idx)) {
+                return undefined;
+              }
+              return React.cloneElement(this.props.seriesComponent, {
```

Comment:
Hmmm... If you expect that the caller could pass a component with pre-set attributes, I think it's OK to use `React.cloneElement`. Otherwise, I would prefer `React.createElement` (https://reactjs.org/docs/react-api.html#createelement).

So, the caller, instead of passing, for example, `seriesComponent={<ChartLine/>}` would pass `seriesComponent={ChartLine}`.

Reply:
Well, the usage of this is not totally defined yet as it's being used internally without passing props but I also expect @aljesusg will use this component as well for tracing and perhaps he'll have to tweak a little bit more the series display, I don't know

Reply:
Since nothing is defined, he even may destroy your code and put something brand new :)